src/kudu/consensus/routing.cc

```diff
@@ -0,0 +1,488 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "kudu/consensus/routing.h"
+
+#include <glog/logging.h>
+#include <google/protobuf/util/message_differencer.h>
+
+#include "kudu/consensus/log_util.h"
+#include "kudu/consensus/quorum_util.h"
+#include "kudu/gutil/map-util.h"
+#include "kudu/gutil/strings/join.h"
+#include "kudu/gutil/strings/substitute.h"
+#include "kudu/util/env.h"
+#include "kudu/util/env_util.h"
+#include "kudu/util/locks.h"
+#include "kudu/util/logging.h"
+#include "kudu/util/pb_util.h"
+#include "kudu/util/scoped_cleanup.h"
+#include "kudu/util/status.h"
+
+using boost::optional;
+using google::protobuf::util::MessageDifferencer;
+using std::shared_ptr;
+using std::unique_ptr;
+using std::string;
+using std::unordered_map;
+using std::vector;
+using strings::Substitute;
+
+namespace kudu {
+namespace consensus {
+
+////////////////////////////////////////////////////////////////////////////////
+// RoutingTable
+////////////////////////////////////////////////////////////////////////////////
+
+Status RoutingTable::Init(const RaftConfigPB& raft_config,
+                          const ProxyTopologyPB& proxy_topology,
+                          const std::string& leader_uuid) {
+  unordered_map<string, Node*> index;
+  unordered_map<string, unique_ptr<Node>> forest;
+
+  Status s = ConstructForest(raft_config, proxy_topology, &index, &forest);
+  if (PREDICT_FALSE(!s.ok() && !s.IsIncomplete())) {
+    return s;
+  }
+  RETURN_NOT_OK(MergeForestIntoSingleRoutingTree(leader_uuid, index, &forest));
+  ConstructNextHopIndicesRec(forest.begin()->second.get());
+
+  index_ = std::move(index);
+  topology_root_ = std::move(forest.begin()->second);
+
+  return s;
+}
+
+Status RoutingTable::ConstructForest(
+    const RaftConfigPB& raft_config,
+    const ProxyTopologyPB& proxy_topology,
+    std::unordered_map<std::string, Node*>* index,
+    std::unordered_map<std::string, std::unique_ptr<Node>>* forest) {
+
+  unordered_map<string, string> dest_to_proxy_from; // keyed by directed edge destination
+  for (const auto& edge : proxy_topology.proxy_edges()) {
+    if (ContainsKey(dest_to_proxy_from, edge.peer_uuid())) {
+      return Status::InvalidArgument("invalid proxy topology: duplicate destination uuid",
+                                     edge.peer_uuid());
+    }
+    InsertOrDie(&dest_to_proxy_from, edge.peer_uuid(), edge.proxy_from_uuid());
+  }
+
+  // Initially, construct a forest comprised of all peers in the Raft config,
+  // with no proxy_from relationships represented.
+  std::unordered_map<std::string, std::unique_ptr<Node>> tmp_forest;
+  std::unordered_map<std::string, Node*> tmp_index;
+  for (const RaftPeerPB& peer : raft_config.peers()) {
+    unique_ptr<Node> hop(new Node(peer));
```

Comment:
nit: you are building a peer to its node mapping in tmp_forest. would this variable be more appropriately just be called node at this point.

Reply:
Ah yes, this was refactored from an edge to a node. Renamed.